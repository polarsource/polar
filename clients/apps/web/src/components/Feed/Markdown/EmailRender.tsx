import Markdown from 'markdown-to-jsx'

import { Container } from '@react-email/components'
import Embed from './Embed/EmailEmbed'
import Iframe from './Iframe/EmailIframe'
import EmailMermaid from './Mermaid/EmailMermaid'
import Paywall, { EmailPaywall } from './Paywall/Paywall'
import Poll from './Poll/Poll'
import EmailSyntaxHighlighter from './SyntaxHighlighter/EmailSyntaxHighlighter'
import {
  RenderArticle,
  markdownOpts,
  wrapStrictCreateElement,
} from './markdown'

export const opts = {
  ...markdownOpts,
  overrides: {
    ...markdownOpts.overrides,

    // email overrides
    poll: (args: any) => <Poll {...args} renderer={EmailPoll} />,
    Paywall: (args: any) => <Paywall {...args} renderer={EmailPaywall} />,
    SubscribeNow: () => <></>, // do not render
    embed: (args: any) => <Embed {...args} />,
    iframe: (args: any) => <Iframe {...args} />,
    pre: (args: any) => {
      if (args.children.type === 'code') {
        const language = args.children.props.className?.replace('lang-', '')
        if (language === 'mermaid') {
          return <EmailMermaid graphDefinition={args.children.props.children} />
        }
        return (
          <EmailSyntaxHighlighter
            language={language}
            {...args.children.props}
          />
        )
      }
      return <></>
    },
  },
} as const

const EmailRender = (props: { article: RenderArticle }) => {
  return (
    <Markdown
      options={{
        ...opts,

        createElement: wrapStrictCreateElement({
          article: props.article,
        }),
      }}
    >
      {props.article.body}
    </Markdown>
  )
}

export default EmailRender

const EmailPoll = (props: { options: string[] }) => {
  return (
    <Container className="my-2 bg-green-300 p-8">
      <table className="w-full">
        {props.options.map((s) => (
          <tr key={s}>
            <td>{s}</td>
            <td className="text-right">123 votes (10%)</td>
          </tr>
        ))}
      </table>
    </Container>
  )
}
