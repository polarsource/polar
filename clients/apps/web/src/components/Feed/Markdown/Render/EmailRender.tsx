import { Highlighter } from '@/components/SyntaxHighlighterShiki/SyntaxHighlighterServer'
import Markdown from 'markdown-to-jsx'

import { Container } from '@react-email/components'
import EmailAd from '../Ad/EmailAd'
import EmailCallout from '../Callout/EmailCallout'
import { calloutRenderRule } from '../Callout/renderRule'
import CodeBlockEmail from '../CodeBlock/CodeBlockEmail'
import Embed from '../Embed/EmailEmbed'
import Iframe from '../Iframe/EmailIframe'
import EmailMermaid from '../Mermaid/EmailMermaid'
import Paywall, { EmailPaywall } from '../Paywall/Paywall'
import Poll from '../Poll/Poll'
import {
  BenefitAds,
  RenderArticle,
  firstChild,
  markdownOpts,
  wrapStrictCreateElement,
} from '../markdown'

export const opts = {
  ...markdownOpts,
  overrides: {
    ...markdownOpts.overrides,

    // email overrides
    poll: (args: any) => <Poll {...args} renderer={EmailPoll} />,
    Paywall: (args: any) => <Paywall {...args} renderer={EmailPaywall} />,
    SubscribeNow: () => <></>, // do not render
    embed: (args: any) => <Embed {...args} />,
    iframe: (args: any) => <Iframe {...args} />,
    Ad: (args: any) => <EmailAd {...args} />,
  },
} as const

const EmailRender = (props: {
  article: RenderArticle
  adsContext?: BenefitAds[]
  highlighter: Highlighter
}) => {
  return (
    <Markdown
      options={{
        ...opts,
        overrides: {
          ...opts.overrides,
          pre: (args: any) => {
            const child = firstChild(args.children)
            if (!child) {
              return <></>
            }

            if (
              typeof child === 'object' &&
              'type' in child &&
              child.type === 'code'
            ) {
              const language = child.props.className?.replace('lang-', '')
              if (language === 'mermaid') {
                const contents = firstChild(child.props.children)
                if (contents && typeof contents === 'string') {
                  return <EmailMermaid graphDefinition={contents} />
                }
              }
              return (
                <CodeBlockEmail
                  language={language}
                  highlighter={props.highlighter}
                  {...child.props}
                />
              )
            }
            return <></>
          },
        },
        createElement: wrapStrictCreateElement({
          article: props.article,
          extraAllowedCustomComponents: Object.keys(opts.overrides),
          adsContext: props.adsContext,
        }),
        renderRule: calloutRenderRule(EmailCallout),
      }}
    >
      {props.article.body}
    </Markdown>
  )
}

export default EmailRender

const EmailPoll = (props: { options: string[] }) => {
  return (
    <Container className="my-2 bg-green-300 p-8">
      <table className="w-full">
        {props.options.map((s) => (
          <tr key={s}>
            <td>{s}</td>
            <td className="text-right">123 votes (10%)</td>
          </tr>
        ))}
      </table>
    </Container>
  )
}
