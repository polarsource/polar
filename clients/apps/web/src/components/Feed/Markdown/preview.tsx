import { Article } from '@polar-sh/sdk'
import { decode } from 'html-entities'
import Markdown from 'markdown-to-jsx'
import { markdownOpts, wrapStrictCreateElement } from './markdown'

export const previewOpts = {
  ...markdownOpts,
  overrides: {
    ...markdownOpts.overrides,
    p: (args: any) => <>{args.children} </>, // Note the space
    div: (args: any) => <>{args.children} </>, // Note the space
    img: (args: any) => <></>,
    a: (args: any) => <>{args.children}</>,
    strong: (args: any) => <>{args.children}</>,
    em: (args: any) => <>{args.children}</>,
  },
} as const

// Text only renderer. Used to generate contents for og:description and similar
export default function PreviewText(props: { article: Article }) {
  return (
    <Markdown
      options={{
        ...previewOpts,
        createElement: wrapStrictCreateElement({
          article: props.article,
          showPaywalledContent: false,
          extraAllowedCustomComponents: [
            'p',
            'div',
            'img',
            'a',
            'strong',
            'em',
          ],

          // Default override that removes all elements
          defaultOverride: (args: any) => <>{args.children}</>,
        }),
        wrapper: (args: any) => <>{args.children}</>,
      }}
    >
      {props.article.body.substring(0, 500)}
    </Markdown>
  )
}

export function UnescapeText(str: string): string {
  return decode(str, { scope: 'strict' }).trim()
}
