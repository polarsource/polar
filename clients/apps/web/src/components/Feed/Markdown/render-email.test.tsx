import { article } from 'polarkit/testdata'

import '@testing-library/jest-dom'
import { render } from '@testing-library/react'
import Markdown from 'markdown-to-jsx'
import { opts } from './EmailRender'
import { RenderArticle, wrapStrictCreateElement } from './markdown'
import { polarPost } from './render.test'
import { polarPostUpsellAccess } from './testdata/polarPostUpsellAccess'

const TestRenderer = (props: { article: RenderArticle }) => {
  return (
    <Markdown
      options={{
        ...opts,
        createElement: wrapStrictCreateElement({
          article: props.article,
          extraAllowedCustomComponents: Object.keys(opts.overrides),
        }),
      }}
    >
      {props.article.body}
    </Markdown>
  )
}

test('code', () => {
  const { container } = render(
    <TestRenderer
      article={{
        ...article,
        body: `
 
code:::

\`\`\`go
func main() {
}
\`\`\`

`,
      }}
    />,
  )
  expect(container).toMatchSnapshot()
})

test('posts', () => {
  for (const post of [polarPost, polarPostUpsellAccess]) {
    const { container } = render(
      <TestRenderer
        article={{
          ...article,
          body: post,
        }}
      />,
    )
    expect(container).toMatchSnapshot()
  }
})
