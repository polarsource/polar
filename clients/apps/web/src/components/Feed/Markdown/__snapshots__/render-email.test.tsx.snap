// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`code 1`] = `
<div>
  <div>
     

    <p>
      code:::
    </p>
    <div
      class="relative my-2 w-full"
    >
      <pre
        style="text-align: left; white-space: pre; word-spacing: normal; word-break: normal; word-wrap: normal; color: rgb(0, 25, 58); background: rgb(243, 244, 247); tab-size: 4; font-size: .8em; border-radius: 2em; hyphens: none; overflow: auto; position: relative; margin: 0.5em 0px; padding: 2em;"
      >
        <code>
          <div
            class="m-0"
          >
            <span
              style="margin-right: 1.5rem; opacity: 0.2; font-size: .7rem; user-select: none; text-align: right; display: inline-block; width: 3.010299956639812px;"
            >
              1
            </span>
            <span
              style="color: rgb(0, 0, 255); font-style: italic;"
            >
              func
            </span>
            <span>
               
            </span>
            <span
              style="color: rgb(124, 77, 255);"
            >
              main
            </span>
            <span
              style="color: rgb(57, 173, 181);"
            >
              (
            </span>
            <span
              style="color: rgb(57, 173, 181);"
            >
              )
            </span>
            <span>
               
            </span>
            <span
              style="color: rgb(57, 173, 181);"
            >
              {
            </span>
          </div>
          <div
            class="m-0"
          >
            <span
              style="margin-right: 1.5rem; opacity: 0.2; font-size: .7rem; user-select: none; text-align: right; display: inline-block; width: 3.010299956639812px;"
            >
              2
            </span>
            <span
              style="color: rgb(57, 173, 181);"
            >
              }
            </span>
          </div>
        </code>
      </pre>
    </div>
  </div>
</div>
`;

exports[`mermaid 1`] = `
<div>
  <div>
    <p>
      we love charts
    </p>
    <center>
      <img
        alt="Mermaid"
        src="http://127.0.0.1:3000/email/kroki/mermaid/eJwryExVKMksyUlVCEgtKVZITMkvKElNUUiqVCjLzynNK0lNLSrmUgACJZf89GIlBSsFYwsziIBzYglYwMIUwg-C8g1NAYd2F64"
      />
    </center>
  </div>
</div>
`;

exports[`posts 1`] = `
<div>
  <div>
    <p>
      <img
        alt="image.png"
        src="https://7vk6rcnylug0u6hg.public.blob.vercel-storage.com/image-fCQqIZt3zubCpaMQzy6ZONZAOG6QDq.png"
        style="max-width: 100%;"
      />
    </p>
    <p>
      Today, we're thrilled to announce & launch the next chapter of Polar.
    </p>
    <p>
      A creator platform for developers and the open source ecosystem â€“Â built open source (
      <a
        href="https://github.com/polarsource/polar"
      >
        fork us on GitHub
      </a>
      ).
    </p>
    <p>
      Offering you â€“ as a developer â€“Â  a platform on top of your GitHub repositories to:
    </p>
    <ol>
      <li>
        Build, own & reach your audience through free- and premium posts and newsletters.
      </li>
      <li>
        Offer subscriptions of value-add benefits designed for our ecosystem & built-in to Polar. We'll also handle value-add taxes so you don't have to.
      </li>
      <li>
        Ability to integrate it all on your own docs, sites or services using our API & SDK.
      </li>
    </ol>
    <p>
      We're also excited to share our updated pricing (it's better for you) and that our team is going to subscribe to open source developers on Polar for $30K in 2024 ðŸŽ‰
    </p>
    <p>
      Let's dive in.
    </p>
    <h3
      id="subscriptions--benefits"
    >
      Subscriptions & Benefits
    </h3>
    <p>
      <img
        alt="Screenshot 2024-01-09 at 10.44.55.png"
        src="https://7vk6rcnylug0u6hg.public.blob.vercel-storage.com/Screenshot%202024-01-09%20at%2010.44.55-uy562dyst8J1Hk4QScE5i6DE3Aa2jA.png"
        style="max-width: 100%;"
      />
    </p>
    <p>
      You can now setup subscriptions with Polar ðŸš€ They're beautiful & come with some incredibly powerful features beyond what you'd expect.
    </p>
    <p>
      <strong>
        Reach your entire audience with free tier(s)
      </strong>
    </p>
    <p>
      You automatically have a built-in - free - tier offering newsletter access to your Public Posts on Polar. Of course, you can edit it to add even more benefits too.
    </p>
    <p>
      Enabling your entire audience to subscribe with email-only at first and for you to have one platform to reach them all.
    </p>
    <p>
      <strong>
        Paid tiers & managed value-add taxes
      </strong>
    </p>
    <p>
      Of course, you can offer paid tiers too that backers can upgrade to seamlessly over time.
    </p>
    <p>
      Many developers shy away from offering paid services & benefits due to the value-add tax implications. Calculating, capturing & remitting value-add taxes internationally is a crushing overhead for individual developers.
    </p>
    <p>
      That's why we're incredibly proud to be the merchant of record and handle value-add taxes so you don't have to. Read more about the details 
      <a
        href="https://docs.polar.sh/payment/"
      >
        here
      </a>
      .
    </p>
    <p>
      <strong>
        Unparalleled business offerings
      </strong>
    </p>
    <p>
      Open source developers tend to have sponsorship tiers for individual users vs. businesses, but they're all treated the same. Not with Polar.
    </p>
    <p>
      Polar has the concept of subscription audiences: Individual vs. Business. Separating them publicly for clarity, but also ensuring benefits are granted to all individual team members of the organization that subscribes ðŸ¤¯
    </p>
    <p>
      Offering unparalleled opportunities and value to subscriptions offered via Polar to businesses.
    </p>
    <p>
      <strong>
        Powerful, built-in, benefits
      </strong>
    </p>
    <p>
      You can setup custom benefits with Polar (free text) for maximum flexibility. However, we're going to invest heavily in automating typical benefits offered within our ecosystem. Making it seamless for you to manage & delightful for your backers to use.
    </p>
    <p>
      From Discord invites (roles/tier) to automating how businesses can setup their logotype & description once and it automatically propagating across your README (PR/SVG) to sites/docs (SDK). Or how about early access to new features, i.e invite to private repositories? Sponsored posts? Just to mention a few ideas in our pipeline.
    </p>
    <p>
      Our first built-in benefit is Posts & Newsletters. Enabling a vital communication channel to an audience you own & can reach fully - regardless of which benefits you offer today or tomorrow.
    </p>
    <h3
      id="posts--newsletters"
    >
      Posts & Newsletters
    </h3>
    <p>
      <img
        alt="image.png"
        src="https://7vk6rcnylug0u6hg.public.blob.vercel-storage.com/image-QKge9Wh9vpdhZRIPLnsMWFesg8SPEY.png"
        style="max-width: 100%;"
      />
    </p>
    <p>
      It's a bit meta, but you're reading a Polar post right now ðŸ˜Ž It's written in markdown, I previewed it easily both on the web & email and I can share it publicly or to paid subscribers.
    </p>
    <p>
      Or how about a hybrid? This is a public post, but the below is a section reserved for paid subscribers. You see that by the upgrade module below (don't worry - I only hid "1337" in there :-)
    </p>
    <p>
      1337
    </p>
    <p>
      We love to write & share code too, so of course we support syntax highlighting in Polar Posts.
    </p>
    <div
      class="relative my-2 w-full"
    >
      <pre
        style="text-align: left; white-space: pre; word-spacing: normal; word-break: normal; word-wrap: normal; color: rgb(0, 25, 58); background: rgb(243, 244, 247); tab-size: 4; font-size: .8em; border-radius: 2em; hyphens: none; overflow: auto; position: relative; margin: 0.5em 0px; padding: 2em;"
      >
        <code>
          <div
            class="m-0"
          >
            <span
              style="margin-right: 1.5rem; opacity: 0.2; font-size: .7rem; user-select: none; text-align: right; display: inline-block; width: 6.989700043360188px;"
            >
              1
            </span>
            <span
              style="color: rgb(0, 0, 255); font-style: italic;"
            >
              def
            </span>
            <span>
               
            </span>
            <span
              style="color: rgb(124, 77, 255);"
            >
              hello_world
            </span>
            <span
              style="color: rgb(57, 173, 181);"
            >
              (
            </span>
            <span
              style="color: rgb(57, 173, 181);"
            >
              )
            </span>
            <span
              style="color: rgb(57, 173, 181);"
            >
              :
            </span>
          </div>
          <div
            class="m-0"
          >
            <span
              style="margin-right: 1.5rem; opacity: 0.2; font-size: .7rem; user-select: none; text-align: right; display: inline-block; width: 6.989700043360188px;"
            >
              2
            </span>
            <span>
                  
            </span>
            <span
              style="color: rgb(0, 0, 255); font-style: italic;"
            >
              print
            </span>
            <span
              style="color: rgb(57, 173, 181);"
            >
              (
            </span>
            <span
              style="color: rgb(239, 100, 100);"
            >
              "Hello world from a Polar Post"
            </span>
            <span
              style="color: rgb(57, 173, 181);"
            >
              )
            </span>
          </div>
          <div
            class="m-0"
          >
            <span
              style="margin-right: 1.5rem; opacity: 0.2; font-size: .7rem; user-select: none; text-align: right; display: inline-block; width: 6.989700043360188px;"
            >
              3
            </span>
            <span />
          </div>
          <div
            class="m-0"
          >
            <span
              style="margin-right: 1.5rem; opacity: 0.2; font-size: .7rem; user-select: none; text-align: right; display: inline-block; width: 6.989700043360188px;"
            >
              4
            </span>
            <span
              style="color: rgb(0, 0, 255); font-style: italic;"
            >
              if
            </span>
            <span>
               __name__ 
            </span>
            <span
              style="color: rgb(57, 173, 181);"
            >
              ==
            </span>
            <span>
               
            </span>
            <span
              style="color: rgb(239, 100, 100);"
            >
              "__main__"
            </span>
            <span
              style="color: rgb(57, 173, 181);"
            >
              :
            </span>
          </div>
          <div
            class="m-0"
          >
            <span
              style="margin-right: 1.5rem; opacity: 0.2; font-size: .7rem; user-select: none; text-align: right; display: inline-block; width: 6.989700043360188px;"
            >
              5
            </span>
            <span>
                  hello_world
            </span>
            <span
              style="color: rgb(57, 173, 181);"
            >
              (
            </span>
            <span
              style="color: rgb(57, 173, 181);"
            >
              )
            </span>
          </div>
        </code>
      </pre>
    </div>
    <p>
      I can embed Youtube videos, images, links or even GitHub issues you can fund using Polar issue funding & rewards.
    </p>
    <table>
      <tbody>
        <tr>
          <td
            class="m-auto"
          >
            <svg
              class="h-4 w-4"
              fill="currentColor"
              viewBox="0 0 20 20"
            >
              <path
                clip-rule="evenodd"
                d="M10 0C4.477 0 0 4.484 0 10.017c0 4.425 2.865 8.18 6.839 9.504.5.092.682-.217.682-.483 0-.237-.008-.868-.013-1.703-2.782.605-3.369-1.343-3.369-1.343-.454-1.158-1.11-1.466-1.11-1.466-.908-.62.069-.608.069-.608 1.003.07 1.531 1.032 1.531 1.032.892 1.53 2.341 1.088 2.91.832.092-.647.35-1.088.636-1.338-2.22-.253-4.555-1.113-4.555-4.951 0-1.093.39-1.988 1.029-2.688-.103-.253-.446-1.272.098-2.65 0 0 .84-.27 2.75 1.026A9.564 9.564 0 0110 4.844c.85.004 1.705.115 2.504.337 1.909-1.296 2.747-1.027 2.747-1.027.546 1.379.203 2.398.1 2.651.64.7 1.028 1.595 1.028 2.688 0 3.848-2.339 4.695-4.566 4.942.359.31.678.921.678 1.856 0 1.338-.012 2.419-.012 2.747 0 .268.18.58.688.482A10.019 10.019 0 0020 10.017C20 4.484 15.522 0 10 0z"
                fill-rule="evenodd"
              />
            </svg>
          </td>
          <td
            class="m-auto pl-2"
          >
            <a
              href="https://github.com/polarsource/polar/issues/897"
            >
              polarsource
              /
              polar
              #
              897
            </a>
          </td>
        </tr>
      </tbody>
    </table>
    <p>
      We've battled with the email gods to make all of this a delightful experience over email too. I can easily choose to share this post over the web only or also email it to all my subscribers (newsletter).
    </p>
    <p>
      Now might be a good time to promote subscribing to our own Polar Posts in the future. So let me embed that nudge easily below.
    </p>
    <p>
      We're working on polls and other embeddables too. We can't wait to hear what you think & how we can improve posts for your initiatives!
    </p>
    <h3
      id="api--sdk"
    >
      API & SDK
    </h3>
    <p>
      You get a personal page on Polar that you can easily promote Posts, Subscriptions & Issue Funding from Day 1: e.g 
      <code>
        polar.sh/polarsource
      </code>
      .
    </p>
    <p>
      We don't think it makes any sense to offer customization abilities to our audience; setting up custom color hex codes, templates etc like other creator platforms. Instead, we're building everything on our public API & offering SDKs (JS first - beta). Since we believe you'd rather integrate it directly on your site, docs and/or services.
    </p>
    <p>
      Our goal is that the developer page, e.g 
      <code>
        polar.sh/polarsource
      </code>
      , should be a living example of an API/SDK integration that you can cherry-pick components from. We'd love to hear from you and your use case and work together to ensure we support it - 
      <a
        href="https://discord.gg/AVgbMhTtBq"
      >
        join our Discord
      </a>
       to chat about it.
    </p>
    <p>
      In the meantime, don't hesitate to checkout our:
    </p>
    <ul>
      <li>
        <a
          href="https://api.polar.sh/redoc"
        >
          Public Polar API
        </a>
      </li>
      <li>
        <a
          href="https://github.com/polarsource/polar/releases/tag/jssdk%2Fv0.2.0"
        >
          JS SDK
        </a>
      </li>
    </ul>
    <p>
      <em>
        It's in early development and we'd love your feedback. We are mindful to maintain backward compatibility, but cannot guarantee it at this stage.
      </em>
    </p>
    <h3
      id="pricing"
    >
      Pricing
    </h3>
    <p>
      We're incredibly excited to share that we're updating our pricing with this launch. In order to drive even more capital to open source developers.
    </p>
    <p>
      Moving forward, our pricing is:
    </p>
    <ul>
      <li>
        No fixed costs
      </li>
      <li>
        5% + payment & payout fees (Stripe) instead of our former 10% + payment & payout fees.
      </li>
      <li>
        <em>
          We'll also cover Stripe fees from our 5% until March 31st, 2024
        </em>
        .
      </li>
    </ul>
    <p>
      Let's go ðŸš€
    </p>
    <h3
      id="30k-in-2024-from-team-polar"
    >
      $30K in 2024 from Team Polar
    </h3>
    <p>
      Finally, we're happy to share that we're going to invest $30K of our own to subscribe to open source developers throughout 2024. So don't hesitate to ping @polar_sh on Twitter with your posts to share it with us and we might subscribe :-)
    </p>
    <p>
      How is this funded?
    </p>
    <ol>
      <li>
        Birk lost it in 2023 and pledged $30K personally and has $20K left to spend.
      </li>
      <li>
        Polar supports businesses to setup open source allowances for team members (invite-only for now). Polar has an employee benefit of each engineer getting $200/month. We're 4 people so that's $800/month or $9,600/year. 
        <em>
          We're rounding up the rest :-)
        </em>
      </li>
    </ol>
    <hr />
    <p>
      Phew. That's all. This is our biggest update since launching Polar and a big step towards our long-term vision of building a platform for developers to get more funding to even building independent businesses around their initiatives.
    </p>
    <p>
      As always, don't hesitate to share your feedback, thoughts or requests on 
      <a
        href="https://github.com/polarsource/polar"
      >
        GitHub
      </a>
       or chat with us directly on 
      <a
        href="https://discord.gg/AVgbMhTtBq"
      >
        Discord
      </a>
      . We're here to serve fellow developers!
    </p>
    <p>
      All our best,
      <br />
      Birk, Gustav, Francois & Emil @ Polar
    </p>
  </div>
</div>
`;

exports[`posts 2`] = `
<div>
  <div>
    <p>
      <img
        alt="Hero image"
        src="https://7vk6rcnylug0u6hg.public.blob.vercel-storage.com/GitHub%20Repo-N4Yx3M434zkSTK3qQHubUyVNwXDohp.png"
        style="max-width: 100%;"
      />
    </p>
    <p>
      You can now seamlessly offer subscribers on Polar access to one, two, three... or countless private GitHub repositories ðŸŽ‰ This opens up and streamlines unlimited possibilities, monetization- and funding models.
    </p>
    <p>
      Open source developers can use it to offer:
    </p>
    <ul>
      <li>
        Access to private GitHub discussions & issues for sponsors
      </li>
      <li>
        Early access to new feature development before upstream push
      </li>
      <li>
        Sponsorware
      </li>
      <li>
        Premium educational materials & code
      </li>
    </ul>
    <p>
      Indie developers can use it to sell:
    </p>
    <ul>
      <li>
        Self-hosting products
      </li>
      <li>
        Courses
      </li>
      <li>
        Starter kits
      </li>
      <li>
        Open core software
      </li>
    </ul>
    <p>
      We're also the merchant of record and take care of VAT.
    </p>
    <p>
      Paired with premium 
      <a
        href="https://polar.sh/polarsource/posts/polar-creator-platform-for-open-source-developers"
      >
        Posts & Newsletters
      </a>
      , 
      <a
        href="https://polar.sh/polarsource/posts/upsell-discord-invites"
      >
        Discord invites
      </a>
       and 
      <a
        href="https://polar.sh/polarsource/posts/automate-sponsor-logos-ads"
      >
        Ads
      </a>
       open source developers can easily offer an unparalleled sponsorship & membership offering. To building an independent commercial open source initiative. You can even build a developer-first SaaS offering.
    </p>
    <p>
      Of course, using our API & SDK all of these capabilities can easily be integrated & upsold on your own site or service too.
    </p>
    <p>
      Polar is built by developers, for developers, and 
      <a
        href="https://github.com/polarsource/polar"
      >
        open source
      </a>
      . We want to empower a new generation of independent developers pursuing their passions full-time. So subscribe to get future product updates & posts as we'll write a series diving deep into all of the opportunities Polar can help facilitate for developers.
    </p>
    <p>
      Or better yet: Start building with Polar today - 
      <a
        href="https://polar.sh"
      >
        signup now
      </a>
      .
    </p>
    <div
      class="bg-gray-200 p-2 text-red-800"
    >
      Invalid &lt;iframe&gt;
    </div>
    <h2
      id="lets-create-a-polar-subscription-benefit-for-a-github-repository"
    >
      Let's create a Polar subscription benefit for a GitHub repository
    </h2>
    <p>
      Let's rig it up and see how it works.
    </p>
    <ol>
      <li>
        Goto 
        <code>
          Subscriptions
        </code>
         in your dashboard and then click 
        <code>
          Edit tier
        </code>
         on your default 
        <strong>
          Free
        </strong>
         tier (for newsletters) to try it out
      </li>
      <li>
        Click 
        <code>
          New Benefit
        </code>
         (See below)
      </li>
    </ol>
    <p>
      <img
        alt="Create new benefit"
        src="https://7vk6rcnylug0u6hg.public.blob.vercel-storage.com/image-ILn7PLAZL2eck6O9rKHCJIO4Nlgd4U.png"
        style="max-width: 100%;"
      />
    </p>
    <ul>
      <li>
        Select 
        <code>
          GitHub Repository Access
        </code>
         as the benefit type
      </li>
      <li>
        Select the 
        <code>
          Organization
        </code>
         the private repository belongs to or connect an organization (See notes below)
      </li>
    </ul>
    <p>
      <strong>
        Why organization vs. personal account?
      </strong>
       GitHub limits personal accounts to only invite collaborators without controlling their permissions, e.g committing to 
      <code>
        main
      </code>
      , pushing new releases etc. Organizations can set more restrictive permissions, e.g read-only. Therefore, due to the security risks, we've put support for personal accounts with this benefit behind a feature flag - reach out if you want it enabled anyway.
    </p>
    <p>
      <strong>
        Important pricing note:
      </strong>
       Collaborators are considered a seat on paid plans and you'll be billed accordingly. We therefore request read-only access to your organizations billing plan to warn about this in our UI so that you can price your Polar subscription to cover it.
    </p>
    <ul>
      <li>
        Click 
        <code>
          Install
        </code>
         to install the required Polar GitHub App
      </li>
    </ul>
    <p>
      <img
        alt="Create benefit - Install GitHub App"
        src="https://7vk6rcnylug0u6hg.public.blob.vercel-storage.com/image-h5l5CsSZ3SI7LBg4Mx5fzokNzjpE6S.png"
        style="max-width: 100%;"
      />
    </p>
    <p>
      You'll be redirected to the GitHub App installation flow to grant access to select repositories. You can choose one of them, a select few or all of them.
    </p>
    <p>
      <strong>
        Important note:
      </strong>
       Selected repositories here are not automatically granted to subscribers on Polar. You'll explicitly select the one (1) desired repository to grant access to for the given subscription benefit you create on Polar. You're in complete control. Of course, you can create multiple benefits to grant access to numerous repositories, but they're all added explicitly.
    </p>
    <p>
      <img
        alt="GitHub installation flow"
        src="https://7vk6rcnylug0u6hg.public.blob.vercel-storage.com/image-0GzjfOkdRqj0QyA07DqFIRstDbsouo.png"
        style="max-width: 100%;"
      />
    </p>
    <p>
      <em>
        Bonus: The Polar GitHub App will also enable you to use the issue crowdfunding & contributor rewards features with Polar too ðŸ”¥. They require read/write to issues & pull requests while administration is required for managing invites to desired GitHub repositories. Unfortunately, GitHub does not offer progressive permissions for Apps.
      </em>
    </p>
    <p>
      Once the Polar GitHub App has been installed you'll be taken back to Polar to continue the setup of the subscription benefit.
    </p>
    <ol>
      <li>
        Select the private 
        <code>
          Repository
        </code>
         you want to grant subscribers access to
      </li>
      <li>
        Select the desired GitHub 
        <code>
          Role
        </code>
        , e.g 
        <code>
          Read
        </code>
         for read access only.
      </li>
      <li>
        Click 
        <code>
          Create
        </code>
      </li>
    </ol>
    <p>
      <img
        alt="Select repository & role"
        src="https://7vk6rcnylug0u6hg.public.blob.vercel-storage.com/image-MenUSRBNEXfw5ZmFkt5uP6Ffsh0I3q.png"
        style="max-width: 100%;"
      />
    </p>
    <p>
      We can now review our updates to the 
      <em>
        Free
      </em>
       tier and see our added GitHub Repository Access benefit we call 
      <code>
        Cutting edge
      </code>
      . We can also update our tier description to celebrate the new benefit.
    </p>
    <p>
      <img
        alt="Review tier"
        src="https://7vk6rcnylug0u6hg.public.blob.vercel-storage.com/image-Z1UfL9ilQFHM64fMqIM3G6vI51bNSv.png"
        style="max-width: 100%;"
      />
    </p>
    <p>
      Looks great. Let's click 
      <code>
        Save Tier
      </code>
       and try it out as a subscriber.
    </p>
    <h2
      id="subscribers-accessing-github-repository"
    >
      Subscribers accessing GitHub repository
    </h2>
    <p>
      <img
        alt="Subscribe"
        src="https://7vk6rcnylug0u6hg.public.blob.vercel-storage.com/image-sGIzRl0if9LmJd9lWhaGDpd6bLnsn3.png"
        style="max-width: 100%;"
      />
    </p>
    <p>
      Your audience can now see the benefit and easily subscribe to gain access to it. In case of email subscriptions, we'll prompt them to connect a GitHub account and immediately create the repository invitation once they have.
    </p>
    <p>
      Subscribers can see & manage their subscription benefits under 
      <code>
        My Subscriptions
      </code>
      .
    </p>
    <p>
      Look at that, they now see the 
      <code>
        Cutting edge
      </code>
       repository with a convenient activation link to it.
    </p>
    <p>
      <em>
        Note: They also receive an email invitation from GitHub.
      </em>
    </p>
    <p>
      <img
        alt="Gain access to repo via subscriptions"
        src="https://7vk6rcnylug0u6hg.public.blob.vercel-storage.com/image-RxC5k4B9QPNxTGwYZUXhsjpjzKxgEw.png"
        style="max-width: 100%;"
      />
    </p>
    <p>
      So let's click the 
      <code>
        Goto polardemo/cutting-edge
      </code>
       link. Since it's our first time, we're prompted to accept the invitation first.
    </p>
    <p>
      <img
        alt="Accept invitation"
        src="https://7vk6rcnylug0u6hg.public.blob.vercel-storage.com/image-qyWQF7WGTkld5JzcbZmvAYjibjspE7.png"
        style="max-width: 100%;"
      />
    </p>
    <p>
      ... and that's it. We're in!
    </p>
    <p>
      <img
        alt="Subscriber seeing the repository on GitHub"
        src="https://7vk6rcnylug0u6hg.public.blob.vercel-storage.com/image-RMwsZSbqgWak7MfKjLjLPD8f25U3jq.png"
        style="max-width: 100%;"
      />
    </p>
    

    <h2
      id="manage-collaborators"
    >
      Manage collaborators
    </h2>
    <p>
      <img
        alt="image.png"
        src="https://7vk6rcnylug0u6hg.public.blob.vercel-storage.com/image-Ihx3RL0EIobi7KlULBBvfeDYleStKA.png"
        style="max-width: 100%;"
      />
    </p>
    <p>
      You can of course review and manage your collaborators on GitHub. As seen above, our subscriber is now a collaborator with 
      <code>
        Read
      </code>
       permissions. Just like we wanted.
    </p>
    <p>
      Oh, and of course, Polar will automatically remove collaborators once their subscription is no longer valid, i.e cancelled.
    </p>
    <hr />
    <p>
      In this example, we've setup seamless invitations to one (1) GitHub repository. However, it's the same process to add additional benefits for more repositories. So there's no limit to how creative you can get with your offerings - we can't wait to see some of your examples!
    </p>
    <p>
      Finally, make sure to subscribe below to stay up-to-date as we ship more product updates & write a series going deeper into different models Polar can help facilitate.
    </p>
    <p>
      Until next time, keep building!
    </p>
  </div>
</div>
`;
